src/plugins/github/cliPlugin.js

```diff
@@ -0,0 +1,108 @@
+// @flow
+
+import Database from "better-sqlite3";
+import fs from "fs-extra";
+import {join as pathJoin} from "path";
+
+import fetchGithubRepo, {fetchGithubRepoFromCache} from "./fetchGithubRepo";
+import type {CacheProvider} from "../../backend/cache";
+import type {CliPlugin, PluginDirectoryContext} from "../../cli2/cliPlugin";
+import type {PluginDeclaration} from "../../analysis/pluginDeclaration";
+import type {ReferenceDetector} from "../../core/references/referenceDetector";
+import type {WeightedGraph} from "../../core/weightedGraph";
+import {Graph} from "../../core/graph";
+import {RelationalView} from "./relationalView";
+import {createGraph} from "./createGraph";
+import {declaration} from "./declaration";
+import {fromRelationalViews as referenceDetectorFromRelationalViews} from "./referenceDetector";
+import {parse as parseConfig, type GithubConfig} from "./config";
+import {validateToken, type GithubToken} from "./token";
+import {weightsForDeclaration} from "../../analysis/pluginDeclaration";
+
+const TOKEN_ENV_VAR_NAME = "SOURCECRED_GITHUB_TOKEN";
+
+async function loadConfig(
+  dirContext: PluginDirectoryContext
+): Promise<GithubConfig> {
+  const dirname = dirContext.configDirectory();
+  const path = pathJoin(dirname, "config.json");
+  const contents = await fs.readFile(path);
+  return Promise.resolve(parseConfig(JSON.parse(contents)));
+}
+
+// Shim to interface with `fetchGithubRepo`; TODO: refactor that to just
+// take a directory.
+class CacheProviderImpl implements CacheProvider {
```

Comment:
Actually, I'm a huge fan of using these interfaces. It decouples the rest of the code from the implementation details of using filesystems. Would recommend applying the same idea to the configuration and tokens too like `ConfigProvider` and `SecretsProvider`.

On the flipside, I would rather not see `fs.readFile` or `process.env` in a _use-case_.

To give a common example where this practice pays off. Many applications use ENV for secrets. However interesting when using Docker Swarm, ENV is not suitable as it will be passed around in plaintext. Instead it should use "Docker Secrets", which most application implement by accepting a `*_FILE` suffix to secrets and reading it from that file path.

Having a central `SecretsProvider` implementation, adding this in would be easy and something the core can implement once. Having `process.env` scattered across the plugins, this would mean every plugin maintainer has to figure this out.

Reply:
> On the flipside, I would rather not see fs.readFile or process.env in a use-case.

I think we're in agreement that stuff like `fs.readFile` and `process.env` should be kept at the outermost layer possible, and decoupled from implementing the logic.

IMO, losing the CacheProvider is not a big deal because we aren't actually supporting other filesystems, so as developers we aren't getting feedback on whether we're (for example) using CacheProvider consistently across the codebase. Without that feedback, there's no particular reason to believe that we *will* actually be using it consistently when we need that flexibility, or even that this is giving us the right flexibility compared to what we'll actually need when we need it. So we don't benefit much from it, and it does add extra indirection and cognitive overhead.

Reply:
I’m not really familiar with the way in which you’re using the term
*use-case*. Perhaps you could clarify?

Currently, `fetchGithubRepo` takes a simple `token: GithubToken`. It’s
up to the caller to fetch that, directly from an environment variable or
from a file or from a cast5-decryption or whatever. It’s beyond the
scope of `fetchGithubRepo` to care where that token came from. I don’t
see how introducing a `secretProvider: () => GithubToken` (presumably?)
makes that any simpler.